explorer-api/src/country_statistics/geolocate.rs

```diff
@@ -0,0 +1,139 @@
+use log::{info, warn};
+use reqwest::Error as ReqwestError;
+use thiserror::Error;
+
+use crate::mix_nodes::{GeoLocation, Location};
+use crate::state::ExplorerApiStateContext;
+
+pub(crate) struct GeoLocateTask {
+    state: ExplorerApiStateContext,
+}
+
+impl GeoLocateTask {
+    pub(crate) fn new(state: ExplorerApiStateContext) -> Self {
+        GeoLocateTask { state }
+    }
+
+    pub(crate) fn start(mut self) {
+        info!("Spawning mix node locator task runner...");
+        tokio::spawn(async move {
+            let mut interval_timer = tokio::time::interval(std::time::Duration::from_millis(50));
+            loop {
+                // wait for the next interval tick
+                interval_timer.tick().await;
+                self.locate_mix_nodes().await;
+            }
+        });
+    }
+
+    async fn locate_mix_nodes(&mut self) {
+        let mixnode_bonds = self.state.inner.mix_nodes.get().await.value;
+
+        for (i, cache_item) in mixnode_bonds.values().enumerate() {
+            if self
+                .state
+                .inner
+                .mix_nodes
+                .is_location_valid(&cache_item.mix_node.identity_key)
+                .await
+            {
+                // when the cached location is valid, don't locate and continue to next mix node
+                continue;
+            }
+
+            // the mix node has not been located or is the cache time has expired
+            match locate(&cache_item.mix_node.host).await {
+                Ok(geo_location) => {
+                    let location = Location::new(geo_location);
+
+                    trace!(
+                        "{} mix nodes already located. Ip {} is located in {:#?}",
+                        i,
+                        cache_item.mix_node.host,
+                        location.three_letter_iso_country_code,
+                    );
+
+                    if i > 0 && (i % 100) == 0 {
+                        info!(
+                            "Located {} mixnodes...",
+                            i + 1,
+                        );
+                    }
+
+                    self.state
+                        .inner
+                        .mix_nodes
+                        .set_location(&cache_item.mix_node.identity_key, Some(location))
+                        .await;
+
+                    // one node has been located, so break out of the loop
+                    return;
+                }
+                Err(e) => match e {
+                    LocateError::ReqwestError(e) => warn!(
+                        "❌ Oh no! Location for {} failed {}",
+                        cache_item.mix_node.host, e
+                    ),
+                    LocateError::NotFound(e) => {
+                            error!(
```

Comment:
Is this really an `error!`? Usually `error` implies some big underlying issue with the program that might cause inevitable crash. If we failed to locate one node, it's not really that big of an issue?